pkg/module/ModuleService.go

```diff
@@ -0,0 +1,180 @@
+ *
+ */
+
+package module
+
+import (
+	"context"
+	"errors"
+	client "github.com/devtron-labs/devtron/api/helm-app"
+	serverEnvConfig "github.com/devtron-labs/devtron/pkg/server/config"
+	util2 "github.com/devtron-labs/devtron/util"
+	"github.com/go-pg/pg"
+	"go.uber.org/zap"
+	"time"
+)
+
+type ModuleService interface {
+	GetModuleInfo(name string) (*ModuleInfoDto, error)
+	HandleModuleAction(userId int32, moduleName string, moduleActionRequest *ModuleActionRequestDto) (*ActionResponse, error)
+}
+
+type ModuleServiceImpl struct {
+	logger                         *zap.SugaredLogger
+	serverEnvConfig                *serverEnvConfig.ServerEnvConfig
+	moduleRepository               ModuleRepository
+	moduleActionAuditLogRepository ModuleActionAuditLogRepository
+	helmAppService                 client.HelmAppService
+	// no need to inject moduleCacheService and cronService, but not generating in wire_gen (not triggering cache work in constructor) if not injecting. hence injecting
+	moduleCacheService ModuleCacheService
+	moduleCronService  ModuleCronService
+}
+
+func NewModuleServiceImpl(logger *zap.SugaredLogger, serverEnvConfig *serverEnvConfig.ServerEnvConfig, moduleRepository ModuleRepository,
+	moduleActionAuditLogRepository ModuleActionAuditLogRepository, helmAppService client.HelmAppService, moduleCacheService ModuleCacheService, moduleCronService ModuleCronService) *ModuleServiceImpl {
+	return &ModuleServiceImpl{
+		logger:                         logger,
+		serverEnvConfig:                serverEnvConfig,
+		moduleRepository:               moduleRepository,
+		moduleActionAuditLogRepository: moduleActionAuditLogRepository,
+		helmAppService:                 helmAppService,
+		moduleCacheService:             moduleCacheService,
+		moduleCronService:              moduleCronService,
+	}
+}
+
+func (impl ModuleServiceImpl) GetModuleInfo(name string) (*ModuleInfoDto, error) {
+	impl.logger.Debugw("getting module info", "name", name)
+
+	moduleInfoDto := &ModuleInfoDto{
+		Name: name,
+	}
+
+	// fetch from DB
+	module, err := impl.moduleRepository.FindOne(name)
+	if err != nil {
+		if err == pg.ErrNoRows {
+			// if entry is not found in database, then treat it as "notInstalled"
+			moduleInfoDto.Status = ModuleStatusNotInstalled
+			return moduleInfoDto, nil
+		}
+		// otherwise some error case
+		impl.logger.Errorw("error in getting module from DB ", "err", err)
+		return nil, err
+	}
+
+	// otherwise send DB status
+	moduleInfoDto.Status = module.Status
+	return moduleInfoDto, nil
+}
+
+func (impl ModuleServiceImpl) HandleModuleAction(userId int32, moduleName string, moduleActionRequest *ModuleActionRequestDto) (*ActionResponse, error) {
+	impl.logger.Debugw("handling module action request", "moduleName", moduleName, "userId", userId, "payload", moduleActionRequest)
+
+	// check if can update server
+	if !impl.serverEnvConfig.CanServerUpdate {
+		return nil, errors.New("module installation is not allowed")
```

Comment:
it will give 500 internal server error, change it to custom api error if possible in this release.

Reply:
Code is tested, would ignore it for now.